spec: abci2.0 -Incorporate content of RFC100 into the spec (#484)
cometbft/cometbft: spec/abci/abci++_comet_expected_behavior.md

  to provide the Application with all the snapshots needed, in order to reconstruct the state locally.
  A successful attempt must provide at least one chunk via `ApplySnapshotChunk`.
  At the end of a successful attempt, CometBFT calls `Info` to make sure the reconstructed state's
  _AppHash_ matches the one in the block header at the corresponding height. Note that the state 
  of  the application does not contain vote extensions itself. The application can rely on 
  [CometBFT to ensure](./../../docs/rfc/rfc-100-abci-vote-extension-propag.md#base-implementation-persist-and-propagate-extended-commit-history)

>   the node has all the relevant data by the time it switches to consensus. 

Reviewer: ```suggestion
  the node has all the relevant data to proceed with the execution beyond this point. 
```
I'm skeptical about using "switching to consensus" in this section, as, from the point of view of the grammar (i.e., the App) consensus is not distinguishable from blocksync.

Author: Ah I reread the sentence and you comment. You are not talking about removing the switching to consensus section, just about rewording this. I agree ,we can change this.

>success-sync        = offer-snapshot 1*apply-chunk
>```

* If vote extensions are enabled, after statesync a node is required to sync at least one block
* via block sync.

<!-- TODO >
Check whether recovery should go after sync-blocks
-->
>```abnf
>block-sync          = 1*sync-blocks recovery
>sync-blocks         = request-block apply-block update-state
>```

Thus the entire sequence of startup via syncing protocols before switching to consensus would look as follows:

>```abnf
>startup          = state-sync block-sync recovery
>```

> 

Reviewer: I would not add this, as the present grammar does not distinguish between block sync and normal consensus behavior. This was done of purpose as the app shouldn't be able to tell.

Finally, `Commit`, which is kept in ABCI++, no longer returns the `AppHash`. It is now up to
`FinalizeBlock` to do so. Thus, a slight refactoring of the old `Commit` implementation will be
needed to move the return of `AppHash` to `FinalizeBlock`.

## Accomodating for vote extensions

In a manner transparent to the application, CometBFT ensures the node is provided with all
the data it needs to participate in consensus. 

In the case of recovering from a crash, or joining the network via state sync, CometBFT will make
sure the node acquires the neccessary vote extensions before switching to conensusus. 

> If a node is already in consensus but falls behind, during catch-up, CometBFT if providing the node with 

Reviewer 2: confusing.
Should `CometBFT if providing` become `CometBFT will provide`?

Author: Yes, spelling mistake.

> vote extensions from past heights by storing the extensions within `ExtendedCommit` for old heights.

Reviewer: ```suggestion
vote extensions from past heights by retrieving the extensions within `ExtendedCommit` for old heights that it had previously stored.
```

We realize this is sub-optimal due to the increase in storage needed to store the extensions, we are 

> working on an optimization of this implementation which should eliviate this concern. The decision to store 

Reviewer: ```suggestion
working on an optimization of this implementation which should alleviate this concern.
However, the application can use the existing `retain_height` parameter to decide how much
history it wants to keep, just as is done with the block history. The network-wide implications
of the usage of `retain_height` stays the same.
The decision to store 
```

historical commits and potential optimizations, are discussed in detail in [RFC-100](./../../docs/rfc/rfc-100-abci-vote-extension-propag.md#current-limitations-and-possible-implementations)

## Handling upgrades to ABCI 2.0 

> If applications upgrade to ABCI 2.0, CometBFT internally ensures that the [application setup](./abci%2B%2B_app_requirements.md#application-configuration-required-to-switch-to-abci-20) is reflected in its opertaion. 

Reviewer: ```suggestion
If applications upgrade to ABCI 2.0, CometBFT internally ensures that the [application setup](./abci%2B%2B_app_requirements.md#application-configuration-required-to-switch-to-abci-20) is reflected in its operation. 
```

Namely, upon saving the block for a given height *h* in the block store at decision time

> - if *h ≥ h<sub>e</sub>*, the corresponding extended commit is saved as  well

Reviewer: You probably need to redefine *h<sub>e</sub>* here (since you can't assume the reader is reading the whole spec)

- if *h < h<sub>e</sub>*, there are no changes to the data saved

> In the catch-up mechanism, when a node f realizes that another peer is at height hp, which is more than 2 heights behind,

Reviewer 2: ```suggestion
In the catch-up mechanism, when a node *f* realizes that another peer is at height *h<sub>p</sub>*, which is more than 2 heights behind,
```

If a node is already in consensus but falls behind, during catch-up, CometBFT will provide the node with 
vote extensions from past heights by retrieving the extensions within `ExtendedCommit` for old heights that it had previously stored.

We realize this is sub-optimal due to the increase in storage needed to store the extensions, we are 
working on an optimization of this implementation which should alleviate this concern.
However, the application can use the existing `retain_height` parameter to decide how much
history it wants to keep, just as is done with the block history. The network-wide implications
of the usage of `retain_height` stay the same.
The decision to store 
historical commits and potential optimizations, are discussed in detail in [RFC-100](./../../docs/rfc/rfc-100-abci-vote-extension-propag.md#current-limitations-and-possible-implementations)

## Handling upgrades to ABCI 2.0 

If applications upgrade to ABCI 2.0, CometBFT internally ensures that the [application setup](./abci%2B%2B_app_requirements.md#application-configuration-required-to-switch-to-abci-20) is reflected in its operation. 

> CometBFT retrieves from the application configuration the value of *h<sub>e</sub>*, the height at which vote extensions are required for consensus to proceed, and uses it to determine the data it stores and data it sends to a peer 

Author: Good idea, done!

Finally, `Commit`, which is kept in ABCI++, no longer returns the `AppHash`. It is now up to
`FinalizeBlock` to do so. Thus, a slight refactoring of the old `Commit` implementation will be
needed to move the return of `AppHash` to `FinalizeBlock`.

## Operational concerns

Introducing vote extensions requires changes to the configuration of the application. Additionally, 
operators and application developers need to be aware of the impact vote extensions have on
expected runtime behaviour.

First of all, switching to a version of CometBFT with vote extensions, requires a coordinated upgrade. 
For a detailed description on the upgrade path, please refer to the corresponding 
[section](./../../docs/rfc/rfc-100-abci-vote-extension-propag.md#upgrade-path) in RFC-100.

> Once upgraded, the configuration file is updated with a [**new consensus parameter**](./abci%2B%2B_app_requirements.md#abciparamsvoteextensionsenableheight): `VoteExtensionsEnableHeight`. 

Reviewer: ```suggestion
Once upgraded, the new version contains a [**new consensus parameter**](./abci%2B%2B_app_requirements.md#abciparamsvoteextensionsenableheight): `VoteExtensionsEnableHeight`. 
```

This parameter represents the height after which vote extensions are 
required for consensus to proceed, with 0 being the default value (no vote extensions).
Once the (coordinated) upgrade to ABCI 2.0 has taken place, at height  *h<sub>u</sub>*,
the value of `VoteExtensionsEnableHeight` MAY be set to some height, *h<sub>e</sub>*,
which MUST be higher than the current height of the chain. Thus the earliest value for 
 *h<sub>e</sub>* is  *h<sub>u</sub>* + 1. 
Once a node reaches the configured height,
for all heights *h ≥ h<sub>e</sub>*, the consensus algorithm will

> reject any votes that do not have vote extension data as invalid.

Reviewer: ```suggestion
reject as invalid any precommit messages that do not have signed vote extension data.
A 0-length vote extensions is allowed, but it MUST be signed.
```

Likewise, for all heights *h < h<sub>e</sub>*, any votes that *do* have vote extensions
will be considered an error condition.
Height *h<sub>e</sub>* is somewhat special, as calls to `PrepareProposal` MUST NOT
have vote extension data, but all precommit votes in that height MUST carry a vote extension.
Height *h<sub>e</sub> + 1* is the first height for which `PrepareProposal` MUST have vote
extension data and all precommit votes in that height MUST have a vote extension.

### Changes to data stored by CometBFT

Upon saving the block for a given height *h* in the block store at decision time

>     - if *h ≥ h<sub>e</sub>*, the corresponding extended commit is saved as  well

Reviewer: ```suggestion
    - if *h ≥ h<sub>e</sub>*, the corresponding extended commit that was used to decide locally is saved as well
```
